src/orderbook.ts

```diff
@@ -103,6 +103,56 @@ export class Orderbook {
     return priceToCoverAmount(new Fraction(amount, 1), asks);
   }
 
+  /**
+   * Removes any overlapping bid/asks which could be matched in the current orderbook
+   * @return A new instance of the orderbook with no more overlapping orders.
+   */
+  reduced() {
+    const result = new Orderbook(this.baseToken, this.quoteToken);
+
+    const bids = Array.from(this.bids.values());
+    bids.sort(sortOffersDescending);
+    const asks = Array.from(this.asks.values());
+    asks.sort(sortOffersAscending);
+
+    const bid_iterator = bids.values();
+    const ask_iterator = asks.values();
+
+    let best_bid = bid_iterator.next();
+    let best_ask = ask_iterator.next();
+    while (
+      !(best_bid.done || best_ask.done) &&
+      !best_bid.value.price.lt(best_ask.value.price)
+    ) {
+      // We have an overlapping bid/ask. Subtract the smaller from the larger and remove the smaller
+      if (best_bid.value.volume.gt(best_ask.value.volume)) {
+        best_bid.value.volume = best_bid.value.volume.sub(
```

Comment:
Does this potentially modify the `this` orderbook's bid object? Again, my JS-foo is not great.

Reply:
Amazing catch. Added a unittest for it and you are right it mutates this :-( Now I'm worried I might have slipped in such mistakes elsewhere in the code.